### PR TITLE
Sharded BlueStore - 1 - KeyValueDB

### PR DESCRIPTION
**Pull request by aclamk**
This is 1/3 of code that allows BlueStore to use RocksDB in sharded mode.
First part does:
- redefines hierarchy of Iterators in KeyValueDB to unify WholeSpaceIterators and Iterators
- changes concept of column families in our RocksDBStore
- allow easy access to values in column families

### CODE REVIEW

**Comment by aclamk** [comment]
Part 1: #29084 
Part 2: #29086
Part 3: #29087

File: src/kv/KeyValueDB.h [hunk 1]
```diff
@@ -204,71 +350,97 @@ class KeyValueDB {
...
-
-  class IteratorImpl : public SimplestIteratorImpl {
-  public:
-    virtual ~IteratorImpl() {}
-    virtual int seek_to_last() = 0;
-    virtual int prev() = 0;
-    virtual std::pair<std::string, std::string> raw_key() = 0;
-    virtual ceph::buffer::ptr value_as_ptr() {
-      ceph::buffer::list bl = value();
+    virtual bufferptr value_as_ptr() {
```
Comment (rzarzynski) [inline src/kv/KeyValueDB.h]: nit: `ceph::bufferptr`.

File: src/kv/RocksDBStore.h [hunk 2]
```diff
@@ -146,48 +152,55 @@ class RocksDBStore : public KeyValueDB {
...
+  int column_family_compact_async(
+		  const std::string& cf_name,
+		  const std::string& prefix,
+		  const std::string& start,
+		  const std::string& end) override;
+
+private:
+  int _open(std::ostream &out, bool read_only, const std::vector<ColumnFamily>& options);
+  std::shared_ptr<rocksdb::MergeOperator> cf_get_merge_operator(const std::string& prefix) const;
+  rocksdb::ColumnFamilyHandle* cf_get_mono_handle(const std::string& cf_name) const;
```
Comment (jdurgin) [inline src/kv/RocksDBStore.h]: would all the callers of cf_get_{mono_}handle() need to take the api_lock as well?

Comment (aclamk) [reply src/kv/RocksDBStore.h]: @jdurgin Right, I fixed all invocations of this internal function to acquire locks.

**Comment by aclamk** [comment]
jenkins test make check

**Comment by vumrao** [comment]
Upstream tracker - https://tracker.ceph.com/issues/41690

**Comment by markhpc** [comment]
retest this please

**Comment by markhpc** [comment]
@aclamk looks like these PRs need a rebase again.

**Comment by aclamk** [comment]
@markhpc Thanks! Rebased.

**Pull request closed without merging**